osu.Game/Screens/Edit/Compose/Components/TriangularPositionSnapGrid.cs

```diff
@@ -0,0 +1,89 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using osu.Framework.Bindables;
+using osu.Game.Utils;
+using osuTK;
+
+namespace osu.Game.Screens.Edit.Compose.Components
+{
+    public partial class TriangularPositionSnapGrid : LinedPositionSnapGrid
+    {
+        /// <summary>
+        /// The spacing between grid lines of this <see cref="TriangularPositionSnapGrid"/>.
+        /// </summary>
+        public BindableFloat Spacing { get; } = new BindableFloat(1f)
+        {
+            MinValue = 0f,
+        };
+
+        /// <summary>
+        /// The rotation in degrees of the grid lines of this <see cref="TriangularPositionSnapGrid"/>.
+        /// </summary>
+        public BindableFloat GridLineRotation { get; } = new BindableFloat();
+
+        public TriangularPositionSnapGrid()
+        {
+            Spacing.BindValueChanged(_ => GridCache.Invalidate());
+            GridLineRotation.BindValueChanged(_ => GridCache.Invalidate());
+        }
+
+        private const float sqrt3 = 1.73205080757f;
+        private const float sqrt3_over2 = 0.86602540378f;
+        private const float one_over_sqrt3 = 0.57735026919f;
```

Comment:
This smells of unnecessary micro-optimisation but I guess I'm willing to let the existence of these consts slide in this particular case since I'm willing to believe it's not - what I am _not_ going to let slide however is the actual hardcoding of these constant values rather than doing the sensible thing of

```suggestion
        private static readonly float sqrt3 = float.Sqrt(3);
        private static readonly float sqrt3_over2 = sqrt3 / 2;
        private static readonly float one_over_sqrt3 = 1 / sqrt3;
```

Reply:
The place where I got the hexgrid algorithm was already micro-optimised so thats why I had these constants.